src/base/notifications/dbusnotifier.h

```diff
@@ -0,0 +1,96 @@
+#include <QVariantMap>
+#include <map>
+
+class OrgFreedesktopNotificationsInterface;
+
+namespace org
+{
+    namespace freedesktop
+    {
+        typedef ::OrgFreedesktopNotificationsInterface Notifications;
+    }
+}
+
+namespace Notifications
+{
+
+    class DBusNotifier: public Notifier
+    {
+        Q_OBJECT
+        typedef Notifier base;
+
+    public:
+        DBusNotifier(QObject* parent);
+        ~DBusNotifier(); // to correctly delete m_notificationsService
+        // Notifier interface
+        virtual void showNotification(const Request &request) const override;
+        void showNotification(Notifications::Request && request) const override;
+
+    private slots:
+        void notificationClosed(uint id, uint reason);
+        void actionInvoked(uint id, const QString &action_key);
+
+    private:
+        enum Capability
+        {
+            ActionIcons = 0x1,      //! Supports using icons instead of text for displaying actions. Using icons for actions must be enabled on a per-notification basis using the "action-icons" hint.
+            Actions = 0x2,          //! The server will provide the specified actions to the user. Even if this cap is missing, actions may still be specified by the client, however the server is free to ignore them.
+            Body = 0x4,             //! Supports body text. Some implementations may only show the summary (for instance, onscreen displays, marquee/scrollers)
+            BodyHyperlinks = 0x8,   //! The server supports hyperlinks in the notifications.
+            BodyImages = 0x10,      //! The server supports images in the notifications.
+            BodyMarkup = 0x20,      //!	Supports markup in the body text. If marked up text is sent to a server that does not give this cap, the markup will show through as regular text so must be stripped clientside.
+            IconMulti = 0x40,       //!	The server will render an animation of all the frames in a given image array. The client may still specify multiple frames even if this cap and/or "icon-static" is missing, however the server is free to ignore them and use only the primary frame.
+            IconStatic = 0x80,      //!	Supports display of exactly 1 frame of any given image array. This value is mutually exclusive with "icon-multi", it is a protocol error for the server to specify both.
+            Persistence = 0x100,    //!	The server supports persistence of notifications. Notifications will be retained until they are acknowledged or removed by the user or recalled by the sender. The presence of this capability allows clients to depend on the server to ensure a notification is seen and eliminate the need for the client to display a reminding function (such as a status icon) of its own.
+            Sound = 0x200,          //! The server supports sounds on notifications. If returned, the server must support the "sound-file" and "suppress-sound" hints.
+        };
+
+		static QStringList getActions(const Request& request);
```

Comment:
???
You definitely forgot to run uncrustify. Then I won't pay attention to this until you do it.